### PR TITLE
feat(string): add binary and octal random string generation

### PR DESCRIPTION
**Pull request by pladreyt**
This PR implements binary and octal random string generation.

Issue: https://github.com/faker-js/faker/issues/184

PR that implements this feature in number module: https://github.com/faker-js/faker/pull/1708

### CODE REVIEW

File: src/modules/string/index.ts [hunk 1]
```diff
@@ -241,6 +241,93 @@ export class StringModule {
     ).join('');
   }
 
+  /**
+   * Returns a [binary](https://en.wikipedia.org/wiki/Binary_number) string.
+   *
+   * @param options The optional options object.
+   * @param options.length The number or range of characters to generate after the prefix. Defaults to `1`.
+   * @param options.prefix Prefix for the generated number. Defaults to `'0x'`.
```
Comment (ST-DDT) [inline src/modules/string/index.ts]: `0x` is the hex prefix IIRC, the binary variant usually use the `0b` variant.

File: src/modules/string/index.ts [hunk 2]
```diff
@@ -241,6 +241,93 @@ export class StringModule {
...
+
+    return `${prefix}${binaryString}`;
+  }
+
+  /**
+   * Returns an [octal](https://en.wikipedia.org/wiki/Octal) string.
+   *
+   * @param options The optional options object.
+   * @param options.length The number or range of characters to generate after the prefix. Defaults to `1`.
+   * @param options.prefix Prefix for the generated number. Defaults to `'0x'`.
```
Comment (ST-DDT) [inline src/modules/string/index.ts]: `0x` is the hex prefix IIRC, the octal variant usually use the `0o` variant.

**Review by ST-DDT — COMMENTED** [review]
Especially octal and binary seem to have a multitude of common prefixes, should we use `''` as the default prefix for all of them?

Only one of our methods uses a non-empty prefix (etheriumAddress), all other 5 methods use ''.

File: src/modules/string/index.ts (hunk 1 above)
Comment (Shinigami92) [reply src/modules/string/index.ts]: Yes, `0x` is actually wrong here! `0b` is the correct form
Same applies for `0x` to `0o` below, ~~but maybe just need to be `0` as prefix in JS~~

~~See:~~



Nevermind: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Deprecated_octal

`0o10 === 8` as example is correct

Comment (pladreyt) [reply src/modules/string/index.ts]: Oh, good catch ! Thanks !. I am fixing it now

File: src/modules/string/index.ts (hunk 2 above)
Comment (pladreyt) [reply src/modules/string/index.ts:288]: Thanks !

**Comment by pladreyt** [comment]
> Especially octal and binary seem to have a multitude of common prefixes, should we use `''` as the default prefix for all of them?
> 
> Only one of our methods uses a non-empty prefix (etheriumAddress), all other 5 methods use ''.

Also the `hexadecimal` method uses a non-empty prefix, I was using it as a base. Was keeping the prefix for consistency with that method, but I can set it to `''`

File: src/modules/string/index.ts [hunk 3]
```diff
@@ -241,6 +241,93 @@ export class StringModule {
...
+   * @param options The optional options object.
+   * @param options.length The number or range of characters to generate after the prefix. Defaults to `1`.
+   * @param options.prefix Prefix for the generated number. Defaults to `'0b'`.
+   *
+   * @example
+   * faker.string.binary() // '0b1'
+   * faker.string.binary({ length: 10 }) // '0b1101011011'
+   * faker.string.binary({ length: { min: 5, max: 10 } }) // '0b11101011'
+   * faker.string.binary({ prefix: '0b' }) // '0b1'
+   * faker.string.binary({ length: 10, prefix: '#' }) // '#1101011011'
```
Comment (Shinigami92) [inline src/modules/string/index.ts]: I think this example doesn't make much sense, but we could swap it out with `bin` for binary 🤔 

```suggestion
   * faker.string.binary({ length: 10, prefix: 'bin' }) // 'bin1101011011'
```

Please let have @ST-DDT also have a look over my suggestion before applying it as I'm not sure

---

Edit: Oh, below in the tests you use already `bin_` and `oct_`!
Please use these here in the examples instead

**Comment by ST-DDT** [comment]
Please add `@see` to and from the number module.

File: src/modules/string/index.ts (hunk 3 above)
Comment (pladreyt) [reply src/modules/string/index.ts]: Sure, sounds good!

**Comment by pladreyt** [comment]
> Please add `@see` to and from the number module.

Sure!

File: src/modules/string/index.ts [hunk 4]
```diff
@@ -241,6 +241,97 @@ export class StringModule {
...
+   * @see faker.number.binary() If you would like to generate a `binary number` (within a range).
+   *
+   * @example
+   * faker.string.binary() // '0b1'
+   * faker.string.binary({ length: 10 }) // '0b1101011011'
+   * faker.string.binary({ length: { min: 5, max: 10 } }) // '0b11101011'
+   * faker.string.binary({ prefix: '0b' }) // '0b1'
+   * faker.string.binary({ length: 10, prefix: 'bin_' }) // 'bin_1101011011'
+   * faker.string.binary({ prefix: '' }) // '1'
+   * faker.string.binary({ length: 10, prefix: '0b' }) // '0b1101011011'
```
Comment (ST-DDT) [inline src/modules/string/index.ts]: These examples are redundant/duplicate (I assume they originally demonstrated the casing property)

File: src/modules/string/index.ts [hunk 5]
```diff
@@ -241,6 +241,97 @@ export class StringModule {
...
+   * @see faker.number.octal() If you would like to generate an `octal number` (within a range).
+   *
+   * @example
+   * faker.string.octal() // '0o3'
+   * faker.string.octal({ length: 10 }) // '0o1526216210'
+   * faker.string.octal({ length: { min: 5, max: 10 } }) // '0o15263214'
+   * faker.string.octal({ prefix: '0o' }) // '0o7'
+   * faker.string.octal({ length: 10, prefix: 'oct_' }) // 'oct_1542153414'
+   * faker.string.octal({ prefix: '' }) // '2'
+   * faker.string.octal({ length: 10, prefix: '0o' }) // '0o1526216210'
```
Comment (ST-DDT) [inline src/modules/string/index.ts]: These examples are redundant/duplicate (I assume they originally demonstrated the casing property)

**Review by ST-DDT — COMMENTED** [review]
Two minor changes, then we are good to go from my side.

File: src/modules/string/index.ts (hunk 4 above)
Comment (pladreyt) [reply src/modules/string/index.ts]: Yes, you are right, will remove them

File: src/modules/string/index.ts (hunk 5 above)
Comment (pladreyt) [reply src/modules/string/index.ts]: Yes, you are right, will remove them

**Review by ST-DDT — APPROVED** [review]
Thanks for your contribution.

**Review by Shinigami92 — APPROVED** [review]

**Pull request merged**